src/transfer/transcode/ome_zarr.py

```diff
@@ -0,0 +1,367 @@
+import fnmatch
+import logging
+import time
+from pathlib import Path
+from typing import Union, List
+
+import dask
+import numpy as np
+import zarr
+from aicsimageio.types import PhysicalPixelSizes
+from aicsimageio.writers import OmeZarrWriter
+from distributed import wait
+from numpy.typing import NDArray
+
+from transfer.util.chunk_utils import (
+    guess_chunks,
+    expand_chunks,
+    ensure_shape_5d,
+    ensure_array_5d,
+)
+from transfer.util.file_utils import collect_filepaths
+from transfer.util.io_utils import (
+    DataReaderFactory,
+    ImarisReader,
+    MissingDatasetError,
+)
+
+logging.basicConfig(format="%(asctime)s %(message)s", datefmt="%Y-%m-%d %H:%M")
+LOGGER = logging.getLogger(__name__)
+LOGGER.setLevel(logging.INFO)
+
+
+def write_files(
+    image_paths: list,
+    output: str,
+    n_levels: int,
+    scale_factor: float,
+    overwrite: bool = True,
+    chunk_size: float = 64,  # MB
+    chunk_shape: tuple = None,
+    voxel_size: tuple = None,
+    storage_options: dict = None,
+) -> list:
+    """
+    Write each image as a separate group to a Zarr store.
+    Each group will be named according to the image filename minus extension.
+    For example, if a file has path /data/tiff/tile_0_0.tif, the group name
+    for that image is tile_0_0. Within each group there are n_levels arrays,
+    one for each resolution level.
+    Args:
+        image_paths: the list of image filepaths to convert to Zarr
+        output: the location of the output Zarr store (filesystem, s3, gs)
+        n_levels: number of downsampling levels
+        scale_factor: scale factor for downsampling in X, Y and Z
+        overwrite: whether to overwrite image groups that already exist
+        chunk_size: the target chunk size in MB
+        chunk_shape: the chunk shape, if None will be computed from chunk_size
+        voxel_size: three element tuple giving physical voxel sizes in Z,Y,X order
+        storage_options: a dictionary of options to pass to the Zarr storage backend, e.g., "compressor"
+    Returns:
+        A list of metrics for each converted image
+    """
+    if chunk_shape is None and chunk_size <= 0:
```

Comment:
yeah, the idea was that at least one of those options must be defined. If `chunk_shape` is passed, it will use that and ignore `chunk_size` regardless of its value. If `chunk_shape` is None, it will use the `chunk_size` to compute the optimal shape. There's probably a cleaner way of doing so.